src/unix/process.c

```diff
@@ -424,20 +426,162 @@ int uv_spawn(uv_loop_t* loop,
 
   /* Acquire write lock to prevent opening new fds in worker threads */
   uv_rwlock_wrlock(&loop->cloexec_lock);
-  pid = fork();
 
-  if (pid == -1) {
-    err = UV__ERR(errno);
-    uv_rwlock_wrunlock(&loop->cloexec_lock);
-    uv__close(signal_pipe[0]);
-    uv__close(signal_pipe[1]);
-    goto error;
-  }
+/* Especial child process spawn case for macOS Big Sur (11.0) onwards 
+ *
+ * Big Sur introduced a significant performance degradation on a call to
+ * fork/exec when the process has many pages mmaped in with MAP_JIT, like, say
+ * a javascript interpreter. Electron-based applications, for example,
+ * are impacted; though the magnitude of the impact depends on how much the
+ * app relies on subprocesses. 
+ * 
+ * On macOS, though, posix_spawn is implemented in a way that does not 
+ * exhibit the problem. This block implements the forking and preparation
+ * logic with poxis_spawn and its related primitves. It also takes advantage of 
+ * the macOS extension POSIX_SPAWN_CLOEXEC_DEFAULT that makes impossible to
+ * leak descriptors to the child process.
+ * 
+ * see https://github.com/libuv/libuv/issues/3050
+ */
+#if defined(__APPLE__)
+  if (__builtin_available(macOS 10.15, *)) {
+    int fd;
+    sigset_t signal_set;
+    posix_spawnattr_t attrs;
+    posix_spawn_file_actions_t actions;
+
+    err = posix_spawnattr_init(&attrs);
+    if(err)
+      goto error;
 
-  if (pid == 0) {
-    uv__process_child_init(options, stdio_count, pipes, signal_pipe[1]);
-    abort();
+    if ((options->flags & UV_PROCESS_SETUID) && posix_spawnattr_set_uid_np(&attrs, options->uid)) {
+      goto error;
+    }
+    if ((options->flags & UV_PROCESS_SETGID) && posix_spawnattr_set_gid_np(&attrs, options->gid)) {
+      goto error;
+    }
+
+    if(options->flags & (UV_PROCESS_SETUID | UV_PROCESS_SETGID)) {
+      /* See the comment on the call to setgroups in uv__process_child_init above
+       * for why this is not a fatal error */
+      SAVE_ERRNO(posix_spawnattr_set_groups_np(&attrs, 0, NULL, KAUTH_UID_NONE));
+    }
+
+    /* Set flags for spawn behavior 
+     * 1) POSIX_SPAWN_CLOEXEC_DEFAULT: (Apple Extension) All descriptors in
+     * the parent will be treated as if they had been created with O_CLOEXEC. 
+     * The only fds that will be passed on to the child are those manipulated by the file actions
+     * 2) POSIX_SPAWN_SETSIGDEF: Signals mentioned in spawn-sigdefault in the spawn attributes
+     * will be reset to behave as their default
+     * 3) POSIX_SPAWN_SETSIGMASK: Signal mask will be set to the value of spawn-sigmask in attributes
+     * 4) POSIX_SPAWN_SETSID: Make the process a new session leader if a detached session was requested. */
+    err = posix_spawnattr_setflags(&attrs, 
+      POSIX_SPAWN_CLOEXEC_DEFAULT |
+      POSIX_SPAWN_SETSIGDEF |
+      POSIX_SPAWN_SETSIGMASK |
+      (options->flags & UV_PROCESS_DETACHED ? POSIX_SPAWN_SETSID : 0)
+      );
```

Comment:
Easier to read when written like this:
```suggestion
    flags = POSIX_SPAWN_CLOEXEC_DEFAULT |
      POSIX_SPAWN_SETSIGDEF |
      POSIX_SPAWN_SETSIGMASK;

    if (options->flags & UV_PROCESS_DETACHED)
      flags |= POSIX_SPAWN_SETSID;

     if (posix_spawnattr_setflags(&attrs, flags))
       goto error;
```